packages/fluentui/react-northstar/src/components/Tree/hooks/flattenTree.ts

```diff
@@ -0,0 +1,144 @@
+import { ObjectShorthandCollection, ShorthandRenderFunction, TreeTitleProps } from '../../../index';
+import { TreeItemProps } from '../TreeItem';
+
+export type BaseFlatTreeItem = {
+  /**
+   * Also in TreeItemProps.
+   * The index of the item among its siblings. Count starts at 1.
+   */
+  index: number;
+
+  /**
+   * Also in TreeItemProps.
+   * Level of the tree/subtree that contains this item.
+   */
+  level: number;
+
+  /**
+   * Also in TreeItemProps.
+   * true if the tree item is expanded (indicating tree item has subtree).
+   * when tree item has no subtree, expanded is undefined
+   */
+  expanded?: boolean;
+
+  /**
+   * Also in TreeItemProps.
+   * parent id of the tree item. For the top level tree items, parent is undefined
+   */
+  parent?: string;
+
+  /**
+   * Also in TreeItemProps.
+   * Size of the tree/subtree that contains this item.
+   */
+  treeSize: number;
+
+  /** true if the tree item has subtree, indicating childrenIds are not undefined */
+  hasSubtree: boolean;
+
+  /** children ids of the tree item. For leaf tree item, childrenIds is undefined */
+  childrenIds?: string[];
+};
+
+export type FlatTree = {
+  [key: string]: BaseFlatTreeItem & { [key: string]: any };
+};
+
+export function flattenTree(
+  items: ObjectShorthandCollection<TreeItemProps>, // tree's props.items
+): FlatTree {
+  const result = {};
+  const flatten = (baseItems, level, parent?) => {
+    if (!baseItems) return;
+
+    if (parent) {
+      result[parent].childrenIds = [];
+    }
+
+    baseItems.forEach((item, indexAmongSiblings, arr) => {
+      const { id, items, expanded } = item;
+      if (parent) {
+        result[parent].childrenIds.push(id);
+      }
+
+      result[id] = {
+        index: indexAmongSiblings + 1, // Used for aria-posinset and it's 1-based.
+        level,
+        expanded: items?.length ? !!expanded : undefined,
+        parent: parent == null ? undefined : parent,
+        treeSize: arr.length,
+        hasSubtree: !!items?.length,
+      };
+
+      if (items?.length) {
+        flatten(items, level + 1, id);
+      }
+    });
+  };
+
+  flatten(items, 1);
+  return result;
+}
```

Comment:
I reworked this function to be more readable and avoid using `.length` as it will always count a length of an array

```suggestion
export function flattenTree(
  items: ObjectShorthandCollection<TreeItemProps>[],
  level: number = 1,
  parent: string | null = null,
  result: BaseFlatTree = {},
): BaseFlatTree {
  const itemsInLeaf = items.length;

  items.forEach((item, indexAmongSiblings) => {
    const { id, expanded, items: childrenItems } = item;
    const hasSubtree = childrenItems ? !!childrenItems.length : false;

    result[id] = {
      index: indexAmongSiblings + 1, // Used for aria-posinset and it's 1-based.
      level,
      expanded: hasSubtree ? !!expanded : undefined,
      parent: parent == null ? undefined : parent,
      treeSize: itemsInLeaf,
      hasSubtree,
      childrenIds: [],
    };

    if (hasSubtree) {
      flattenTree(childrenItems, level + 1, id, result);
    }

    if (parent) {
      result[parent].childrenIds.push(id);
    }
  });

  return result;
}
```

A test on CSB to ensure that implementation matches: https://codesandbox.io/s/throbbing-dew-ewbcp?file=/src/flatten.spec.js:852-1603

The single difference is that it always assigns `childrenItems`. Perf is the same:

![image](https://user-images.githubusercontent.com/14183168/98233540-82b22800-1f5f-11eb-9032-15c53d9eb884.png)

Reply:
I added a check to make sure `childrenIds` is undefined instead of `[]` for leaf nodes. I discover I was checking if an item is leaf by if childrenIds is null. So I'd like to keep this.